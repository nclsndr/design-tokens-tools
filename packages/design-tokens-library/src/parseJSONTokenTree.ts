import { buildTreeState } from './state/buildTreeState.js';
import { TokenTree } from './client/TokenTree.js';

/**
 * Parse a JSON token tree into a TokenTree instance.
 * @param jsonTokenTree - The JSON token tree to parse, must be string or object.
 * @param options
 * @param options.name - The name of the token tree.
 * @param options.throwOnError - Throw an error if the token tree is invalid.
 */
export function parseJSONTokenTree(
  jsonTokenTree: unknown,
  options?: { name?: string; throwOnError?: boolean },
) {
  const treeState = buildTreeState(jsonTokenTree, options);
  return new TokenTree(treeState);
}
