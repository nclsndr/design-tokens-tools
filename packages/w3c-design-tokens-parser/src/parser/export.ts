export { captureAliasPath } from './alias/captureAliasPath.js';
export { parseAliasValue } from './alias/parseAliasValue.js';

export type { AnalyzedGroup } from './group/AnalyzedGroup.js';
export { parseRawGroup } from './group/parseRawGroup.js';

export { type AnalyzedValue, AnalyzedToken } from './token/AnalyzedToken.js';
export { captureAnalyzedTokensReferenceErrors } from './token/captureAnalyzedTokensReferenceErrors.js';
export { findAnalyzedTokenByPath } from './token/findAnalyzedTokenByPath.js';
export { matchTokenTypeAgainstAliasingMapping } from './token/matchTokenTypeAgainstAliasingMapping.js';
export { parseRawToken } from './token/parseRawToken.js';
export { parseTokenTypeName } from './token/parseTokenTypeName.js';
export {
  type ReferenceResolutionTrace,
  type LinkedReferenceResolutionTrace,
  type UnlinkedReferenceResolutionTrace,
  recursivelyResolveAnalyzedToken,
} from './token/recursivelyResolveAnalyzedToken.js';
export {
  type ResolutionType,
  recursivelyResolveTokenType,
} from './token/recursivelyResolveTokenType.js';

export { parseRawInput } from './tree/parseRawInput.js';
export { parseTreeNode } from './tree/parseTreeNode.js';
export { parseTreeNodeDescription } from './tree/parseTreeNodeDescription.js';
export { parseTreeNodeExtensions } from './tree/parseTreeNodeExtensions.js';

export type { AnalyzerContext } from './utils/AnalyzerContext.js';
