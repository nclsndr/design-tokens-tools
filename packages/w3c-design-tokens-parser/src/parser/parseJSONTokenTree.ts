import { Cause, Effect, Exit } from 'effect';
import {
  matchIsToken,
  matchIsGroup,
  ALIAS_PATH_SEPARATOR,
  JSON as JSONTypes,
} from 'design-tokens-format-module';

import { traverseJSONValue } from '../utils/traverseJSONValue.js';
import { parseRawToken } from './token/parseRawToken.js';
import { ValidationError } from '../utils/validationError.js';
import { parseRawGroup } from './group/parseRawGroup.js';
import { AnalyzedToken } from './token/AnalyzedToken.js';
import { AnalyzedGroup } from './group/AnalyzedGroup.js';
import { parseTreeNode } from './tree/parseTreeNode.js';
import { makeUniqueId } from '../utils/uniqueId.js';
import { AnalyzerContext } from './utils/AnalyzerContext.js';

function endsWith(arr: Array<string | number>, end: string | number): boolean {
  if (arr.length === 0) return false;
  return arr[arr.length - 1] === end;
}

function parseRawInput(
  input: unknown,
  ctx: AnalyzerContext,
): Effect.Effect<JSONTypes.Object, Array<ValidationError>, never> {
  if (typeof input === 'string') {
    return Effect.try({
      try: () => JSON.parse(input) as JSONTypes.Object,
      catch: () => {
        return [
          new ValidationError({
            nodeId: '',
            type: 'Computation',
            message: 'Failed to parse JSON string',
            treePath: [],
          }),
        ];
      },
    });
  }
  return parseTreeNode(input, ctx);
}

export const parseJSONTokenTree: (input: unknown) => Effect.Effect<
  {
    tokenTree: JSONTypes.Object;
    analyzedTokens: Array<AnalyzedToken>;
    analyzedGroups: Array<AnalyzedGroup>;
    tokenErrors: Array<ValidationError>;
    groupErrors: Array<ValidationError>;
  },
  ValidationError[],
  never
> = (input) =>
  parseRawInput(input, {
    varName: '[root]',
    nodeId: '',
    path: [],
  }).pipe(
    Effect.flatMap((jsonTokenTree) => {
      const analyzedTokenEffects: Array<
        Effect.Effect<AnalyzedToken, Array<ValidationError>>
      > = [];
      const analyzedGroupEffects: Array<
        Effect.Effect<AnalyzedGroup, Array<ValidationError>>
      > = [];

      traverseJSONValue(jsonTokenTree, (value, rawPath) => {
        if (
          endsWith(rawPath, '$description') ||
          endsWith(rawPath, '$extensions')
        ) {
          return false;
        }

        const nodeId = makeUniqueId();

        if (matchIsToken(value)) {
          analyzedTokenEffects.push(
            parseRawToken(value, {
              jsonTokenTree,
              nodeId: nodeId,
              path: rawPath,
              varName: `${rawPath.join(ALIAS_PATH_SEPARATOR)}`,
            }),
          );

          return false;
        } else if (matchIsGroup(value)) {
          analyzedGroupEffects.push(
            parseRawGroup(value, {
              nodeId: nodeId,
              path: rawPath,
              varName: `${rawPath.join(ALIAS_PATH_SEPARATOR)}`,
            }),
          );

          return true;
        }
        return false;
      });

      const tokenErrors: Array<ValidationError> = [];
      const analyzedTokens: Array<AnalyzedToken> = [];
      for (const tokenEffect of analyzedTokenEffects) {
        Exit.match(Effect.runSyncExit(tokenEffect), {
          onSuccess: (value) => {
            analyzedTokens.push(value);
          },
          onFailure: (cause) => {
            Cause.match(cause, {
              onEmpty: undefined,
              onFail: (errors) => {
                tokenErrors.push(...errors);
              },
              onDie: () => {},
              onInterrupt: () => {},
              onSequential: () => {},
              onParallel: () => {},
            });
          },
        });
      }

      const groupErrors: Array<ValidationError> = [];
      const analyzedGroups: Array<AnalyzedGroup> = [];

      for (const groupEffect of analyzedGroupEffects) {
        Exit.match(Effect.runSyncExit(groupEffect), {
          onSuccess: (value) => {
            analyzedGroups.push(value);
          },
          onFailure: (cause) => {
            Cause.match(cause, {
              onEmpty: undefined,
              onFail: (errors) => {
                groupErrors.push(...errors);
              },
              onDie: () => {},
              onInterrupt: () => {},
              onSequential: () => {},
              onParallel: () => {},
            });
          },
        });
      }

      return Effect.succeed({
        tokenTree: jsonTokenTree,
        analyzedTokens,
        analyzedGroups,
        tokenErrors,
        groupErrors,
      });
    }),
  );
