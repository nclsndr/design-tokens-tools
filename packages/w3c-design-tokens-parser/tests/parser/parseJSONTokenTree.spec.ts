import { describe, expect, it } from 'vitest';
import { Either, Option } from 'effect';
import { JSONTokenTree } from 'design-tokens-format-module';

import {
  borderToken,
  colorToken,
  cubicBezierToken,
  dimensionToken,
  durationToken,
  fontFamilyToken,
  gradientToken,
  numberFontWeightToken,
  numberToken,
  shadowToken,
  stringFontWeightToken,
  strokeStyleToken,
  transitionToken,
  typographyToken,
} from '../_fixtures/tokens';

import { parseJSONTokenTree } from '../../src/parser/parseJSONTokenTree';

describe('parseJSONTokenTree', () => {
  it('should parse a token tree embedded in a string', () => {
    const tokens: JSONTokenTree = {
      base: {
        primary: colorToken,
      },
      semantic: {
        primary: {
          $value: '{base.primary}',
        },
      },
    };

    const result = parseJSONTokenTree(JSON.stringify(tokens));

    const matched = Either.getOrThrow(result);

    expect(matched?.analyzedTokens).toHaveLength(2);
    expect(matched?.analyzedGroups).toHaveLength(3);
    expect(matched?.tokenErrors).toStrictEqual([]);
    expect(matched?.groupErrors).toStrictEqual([]);
  });
  it('should parse a token tree of raw values of all types', () => {
    const tree: JSONTokenTree = {
      borderToken,
      colorToken,
      cubicBezierToken,
      dimensionToken,
      durationToken,
      fontFamilyToken,
      stringFontWeightToken,
      numberFontWeightToken,
      gradientToken,
      numberToken,
      shadowToken,
      strokeStyleToken,
      transitionToken,
      typographyToken,
    };

    const result = parseJSONTokenTree(tree);

    const matched = Either.match(result, {
      onRight: (r) => ({
        ...r,
        analyzedTokens: r.analyzedTokens.map((t) => ({
          ...JSON.parse(JSON.stringify(t)),
          id: 'ID',
        })),
      }),
      onLeft: () => undefined,
    });

    expect(matched?.analyzedTokens).toHaveLength(14);
    expect(matched?.analyzedTokens).toMatchSnapshot();
  });
  it('should parse a token tree with $description and $extensions at the root level', () => {
    const tree: JSONTokenTree = {
      $description: 'A description for the token tree.',
      $extensions: {
        someExtension: 'Some value',
      },
      base: {
        primary: colorToken,
      },
    };

    const result = parseJSONTokenTree(tree);

    const matched = JSON.parse(JSON.stringify(Either.getOrThrow(result)));

    expect(matched).toStrictEqual({
      tokenTree: {
        $description: 'A description for the token tree.',
        $extensions: {
          someExtension: 'Some value',
        },
        base: {
          primary: {
            $type: 'color',
            $value: '#a82222',
          },
        },
      },
      analyzedTokens: [
        {
          id: expect.any(String),
          path: ['base', 'primary'],
          type: 'color',
          value: {
            raw: '#a82222',
            toReferences: [],
          },
        },
      ],
      analyzedGroups: [
        {
          id: expect.any(String),
          path: [],
          childrenCount: 1,
          description: 'A description for the token tree.',
          extensions: {
            someExtension: 'Some value',
          },
        },
        {
          id: expect.any(String),
          path: ['base'],
          childrenCount: 1,
        },
      ],
      tokenErrors: [],
      groupErrors: [],
    });
  });
  it('should parse a token tree with a token at the top level', () => {
    const tree: JSONTokenTree = colorToken;

    const result = parseJSONTokenTree(tree);

    const matched = JSON.parse(JSON.stringify(Either.getOrThrow(result)));

    expect(matched).toStrictEqual({
      tokenTree: {
        $type: 'color',
        $value: '#a82222',
      },
      analyzedTokens: [
        {
          id: expect.any(String),
          path: [],
          type: 'color',
          value: {
            raw: '#a82222',
            toReferences: [],
          },
        },
      ],
      analyzedGroups: [],
      tokenErrors: [],
      groupErrors: [],
    });
  });
  it('should parse a token tree with references to other tokens', () => {
    const tree: JSONTokenTree = {
      base: {
        primary: colorToken,
      },
      semantic: {
        primary: {
          $value: '{base.primary}',
        },
      },
    };

    const result = parseJSONTokenTree(tree);

    const matched = JSON.parse(JSON.stringify(Either.getOrThrow(result)));

    expect(matched).toStrictEqual({
      tokenTree: {
        base: {
          primary: {
            $type: 'color',
            $value: '#a82222',
          },
        },
        semantic: {
          primary: {
            $value: '{base.primary}',
          },
        },
      },
      analyzedTokens: [
        {
          id: expect.any(String),
          path: ['base', 'primary'],
          type: 'color',
          value: {
            raw: '#a82222',
            toReferences: [],
          },
        },
        {
          id: expect.any(String),
          path: ['semantic', 'primary'],
          type: 'color',
          value: {
            raw: '{base.primary}',
            toReferences: [
              {
                fromTreePath: ['semantic', 'primary'],
                fromValuePath: [],
                toTreePath: ['base', 'primary'],
              },
            ],
          },
        },
      ],
      analyzedGroups: [
        {
          id: expect.any(String),
          path: [],
          childrenCount: 2,
        },
        {
          id: expect.any(String),
          path: ['base'],
          childrenCount: 1,
        },
        {
          id: expect.any(String),
          path: ['semantic'],
          childrenCount: 1,
        },
      ],
      tokenErrors: [],
      groupErrors: [],
    });
  });
  it('should parse a token tree with invalid tokens', () => {
    const tree: JSONTokenTree = {
      base: {
        // @ts-expect-error
        primary: {
          $type: 'color',
          $value: 'rgba(0, 0, 0, 0.5)',
        },
      },
      semantic: {
        primary: {
          // @ts-expect-error
          $type: 'unknown',
          // @ts-expect-error
          $value: 'some value',
        },
      },
      invalidGroup: {
        // @ts-expect-error
        $description: 42,
      },
    };

    const result = parseJSONTokenTree(tree);

    const matched = JSON.parse(JSON.stringify(Either.getOrThrow(result)));

    expect(matched).toStrictEqual({
      tokenTree: {
        base: { primary: { $type: 'color', $value: 'rgba(0, 0, 0, 0.5)' } },
        semantic: { primary: { $type: 'unknown', $value: 'some value' } },
        invalidGroup: { $description: 42 },
      },
      analyzedTokens: [],
      analyzedGroups: [
        { id: expect.any(String), path: [], childrenCount: 3 },
        { id: expect.any(String), path: ['base'], childrenCount: 1 },
        { id: expect.any(String), path: ['semantic'], childrenCount: 1 },
      ],
      tokenErrors: [
        {
          type: 'Value',
          isCritical: false,
          nodeId: expect.any(String),
          treePath: ['base', 'primary'],
          nodeKey: '$value',
          valuePath: [],
          message:
            'base.primary.$value must start with "#" and have a length of 6 or 8. Got: "rgba(0, 0, 0, 0.5)".',
        },
        {
          type: 'Value',
          isCritical: false,
          nodeId: expect.any(String),
          treePath: ['semantic', 'primary'],
          nodeKey: '$type',
          valuePath: [],
          message:
            'semantic.primary.$type must be a valid type among: "color", "dimension", "fontFamily", "fontWeight", "duration", "cubicBezier", "number", "strokeStyle", "border", "transition", "shadow", "gradient", "typography". Got "unknown".',
        },
      ],
      groupErrors: [
        {
          type: 'Type',
          isCritical: false,
          nodeId: expect.any(String),
          treePath: ['invalidGroup'],
          nodeKey: '$description',
          valuePath: [],
          message: 'invalidGroup.$description must be a string. Got "number".',
        },
      ],
    });
  });
  it('should parse a token tree containing a circular reference', () => {
    const tree: JSONTokenTree = {
      base: {
        $type: 'color',
        primary: {
          $value: '{base.primary}',
        },
      },
    };

    const result = parseJSONTokenTree(tree);

    const matched = JSON.parse(JSON.stringify(Either.getOrThrow(result)));

    expect(matched).toStrictEqual({
      tokenTree: {
        base: { $type: 'color', primary: { $value: '{base.primary}' } },
      },
      analyzedTokens: [],
      analyzedGroups: [
        { id: expect.any(String), path: [], childrenCount: 1 },
        {
          id: expect.any(String),
          path: ['base'],
          tokenType: 'color',
          childrenCount: 1,
        },
      ],
      tokenErrors: [
        {
          isCritical: false,
          message:
            'Circular references detected while resolving token type for token "base.primary".',
          referenceToTreePath: ['base', 'primary'],
          type: 'Computation',
          nodeId: '',
          treePath: ['base', 'primary'],
          valuePath: [],
        },
      ],
      groupErrors: [],
    });
  });
});
