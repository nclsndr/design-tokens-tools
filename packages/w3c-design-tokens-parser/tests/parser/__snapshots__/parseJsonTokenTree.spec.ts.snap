// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`parseJSONTokenTree > should parse a token tree of raw values of all types 1`] = `
[
  {
    "id": "ID",
    "path": [
      "borderToken",
    ],
    "type": "border",
    "value": {
      "raw": {
        "color": "#676767",
        "style": "solid",
        "width": "1px",
      },
      "toReferences": [],
    },
  },
  {
    "id": "ID",
    "path": [
      "colorToken",
    ],
    "type": "color",
    "value": {
      "raw": "#a82222",
      "toReferences": [],
    },
  },
  {
    "id": "ID",
    "path": [
      "cubicBezierToken",
    ],
    "type": "cubicBezier",
    "value": {
      "raw": [
        0,
        1,
        1,
        0,
      ],
      "toReferences": [],
    },
  },
  {
    "id": "ID",
    "path": [
      "dimensionToken",
    ],
    "type": "dimension",
    "value": {
      "raw": "12px",
      "toReferences": [],
    },
  },
  {
    "id": "ID",
    "path": [
      "durationToken",
    ],
    "type": "duration",
    "value": {
      "raw": "1ms",
      "toReferences": [],
    },
  },
  {
    "id": "ID",
    "path": [
      "fontFamilyToken",
    ],
    "type": "fontFamily",
    "value": {
      "raw": "Arial",
      "toReferences": [],
    },
  },
  {
    "id": "ID",
    "path": [
      "stringFontWeightToken",
    ],
    "type": "fontWeight",
    "value": {
      "raw": "bold",
      "toReferences": [],
    },
  },
  {
    "id": "ID",
    "path": [
      "numberFontWeightToken",
    ],
    "type": "fontWeight",
    "value": {
      "raw": 700,
      "toReferences": [],
    },
  },
  {
    "id": "ID",
    "path": [
      "gradientToken",
    ],
    "type": "gradient",
    "value": {
      "raw": [
        {
          "color": "#000000",
          "position": 0,
        },
        {
          "color": "#ffffff",
          "position": 1,
        },
      ],
      "toReferences": [],
    },
  },
  {
    "id": "ID",
    "path": [
      "numberToken",
    ],
    "type": "number",
    "value": {
      "raw": 12,
      "toReferences": [],
    },
  },
  {
    "id": "ID",
    "path": [
      "shadowToken",
    ],
    "type": "shadow",
    "value": {
      "raw": {
        "blur": "3px",
        "color": "#000000",
        "offsetX": "1px",
        "offsetY": "2px",
        "spread": "4px",
      },
      "toReferences": [],
    },
  },
  {
    "id": "ID",
    "path": [
      "strokeStyleToken",
    ],
    "type": "strokeStyle",
    "value": {
      "raw": "solid",
      "toReferences": [],
    },
  },
  {
    "id": "ID",
    "path": [
      "transitionToken",
    ],
    "type": "transition",
    "value": {
      "raw": {
        "delay": "0ms",
        "duration": "100ms",
        "timingFunction": [
          0,
          0.1,
          0.7,
          0.5,
        ],
      },
      "toReferences": [],
    },
  },
  {
    "id": "ID",
    "path": [
      "typographyToken",
    ],
    "type": "typography",
    "value": {
      "raw": {
        "fontFamily": "Arial",
        "fontSize": "12px",
        "fontWeight": "bold",
        "letterSpacing": "1px",
        "lineHeight": 1.5,
      },
      "toReferences": [],
    },
  },
]
`;
